From 1e17457b99a2f60ae6762a33e2f7f3c1e2f3bcbb Mon Sep 17 00:00:00 2001
From: Chris Friesen <chris.friesen@windriver.com>
Date: Tue, 24 Nov 2015 16:27:29 -0500
Subject: [PATCH] Affine irqs and workqueues with kthread_cpus

If the kthread_cpus boot arg is set it means we want to affine
kernel threads to the specified CPU mask as much as possible
in order to avoid doing work on other CPUs.

In this commit we extend the meaning of that boot arg to also
apply to the CPU affinity of unbound and ordered workqueues.

We also use the kthread_cpus value to determine the default irq
affinity.  Specifically, as long as the previously-calculated
irq affinity intersects with the kthread_cpus affinity then we'll
use the intersection of the two as the default irq affinity.

Signed-off-by: Chris Friesen <chris.friesen@windriver.com>
[VT: replacing spaces with tabs.  Performed tests]
Signed-off-by: Vu Tran <vu.tran@windriver.com>

Signed-off-by: Jim Somerville <Jim.Somerville@windriver.com>
Signed-off-by: Zhang Zhiguo <zhangzhg@neusoft.com>
Signed-off-by: Jiping Ma <jiping.ma2@windriver.com>
[lz: Adapted the patch for context changes.]
Signed-off-by: Li Zhou <li.zhou@windriver.com>
---
 kernel/irq/manage.c | 7 +++++++
 kernel/workqueue.c  | 4 ++++
 2 files changed, 11 insertions(+)

diff --git a/kernel/irq/manage.c b/kernel/irq/manage.c
index d309ba84e..6965ad7d0 100644
--- a/kernel/irq/manage.c
+++ b/kernel/irq/manage.c
@@ -627,6 +627,13 @@ int irq_setup_affinity(struct irq_desc *desc)
 		if (cpumask_intersects(&mask, nodemask))
 			cpumask_and(&mask, &mask, nodemask);
 	}
+
+	/* This will narrow down the affinity further if we've specified
+	 * a reduced cpu_kthread_mask in the boot args.
+	 */
+	if (cpumask_intersects(&mask, cpu_kthread_mask))
+		cpumask_and(&mask, &mask, cpu_kthread_mask);
+
 	ret = irq_do_set_affinity(&desc->irq_data, &mask, false);
 	raw_spin_unlock(&mask_lock);
 	return ret;
diff --git a/kernel/workqueue.c b/kernel/workqueue.c
index 3a2cbe9e1..82246a3e5 100644
--- a/kernel/workqueue.c
+++ b/kernel/workqueue.c
@@ -6591,6 +6591,8 @@ void __init workqueue_init_early(void)
 
 		BUG_ON(!(attrs = alloc_workqueue_attrs()));
 		attrs->nice = std_nice[i];
+		/* If we've specified a kthread mask apply it here too. */
+		cpumask_copy(attrs->cpumask, cpu_kthread_mask);
 		unbound_std_wq_attrs[i] = attrs;
 
 		/*
@@ -6600,6 +6602,8 @@ void __init workqueue_init_early(void)
 		BUG_ON(!(attrs = alloc_workqueue_attrs()));
 		attrs->nice = std_nice[i];
 		attrs->ordered = true;
+		/* If we've specified a kthread mask apply it here too. */
+		cpumask_copy(attrs->cpumask, cpu_kthread_mask);
 		ordered_wq_attrs[i] = attrs;
 	}
 
-- 
2.17.1

